drivers/serial/usart_gd32.c

```diff
@@ -124,27 +146,189 @@ static int usart_gd32_err_check(const struct device *dev)
 	return errors;
 }
 
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+int usart_gd32_fifo_fill(const struct device *dev, const uint8_t *tx_data,
+			 int len)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+	uint8_t num_tx = 0U;
+
+	while ((len - num_tx > 0) &&
+		usart_flag_get(cfg->reg, USART_FLAG_TBE)) {
+		usart_data_transmit(cfg->reg, tx_data[num_tx++]);
+	}
+
+	return num_tx;
+}
+
+int usart_gd32_fifo_read(const struct device *dev, uint8_t *rx_data,
+			 const int size)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+	uint8_t num_rx = 0U;
+
+	while ((size - num_rx > 0) &&
+		usart_flag_get(cfg->reg, USART_FLAG_RBNE)) {
+		rx_data[num_rx++] = usart_data_receive(cfg->reg);
+	}
+
+	return num_rx;
+}
+
+void usart_gd32_irq_tx_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_TC);
+}
+
+void usart_gd32_irq_tx_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_TC);
+}
+
+int usart_gd32_irq_tx_ready(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_TBE) &&
+	       usart_interrupt_flag_get(cfg->reg, USART_INT_FLAG_TC);
+}
+
+int usart_gd32_irq_tx_complete(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_TC);
+}
+
+void usart_gd32_irq_rx_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_RBNE);
+}
+
+void usart_gd32_irq_rx_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_RBNE);
+}
+
+int usart_gd32_irq_rx_ready(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_RBNE);
+}
+
+void usart_gd32_irq_err_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_ERR);
+	usart_interrupt_enable(cfg->reg, USART_INT_PERR);
+}
+
+void usart_gd32_irq_err_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_ERR);
+	usart_interrupt_disable(cfg->reg, USART_INT_PERR);
+}
+
+int usart_gd32_irq_is_pending(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return ((usart_flag_get(cfg->reg, USART_FLAG_RBNE) &&
+		 usart_interrupt_flag_get(cfg->reg, USART_INT_FLAG_RBNE)) ||
+		(usart_flag_get(cfg->reg, USART_FLAG_TC) &&
+		 usart_interrupt_flag_get(cfg->reg, USART_INT_FLAG_TC)));
+}
+
+int usart_gd32_irq_update(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return 1;
+}
```

Comment:
it is not necessary to implement this one (see uart.h, returns `-ENOSYS` if NULL)

Reply:
removed.